vms/platformvm/state/iteration.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package state
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/hashing"
+)
+
+const (
+	startWeightKeyLength  = hashing.HashLen + database.Uint64Size
+	weightKeyLength       = startWeightKeyLength + hashing.AddrLen
+	weightKeyHeightOffset = hashing.HashLen
+	weightKeyNodeIDOffset = weightKeyHeightOffset + database.Uint64Size
+
+	startBLSKeyLength  = database.Uint64Size
+	blsKeyLength       = startBLSKeyLength + hashing.AddrLen
+	blsKeyNodeIDOffset = database.Uint64Size
+)
+
+var (
+	errUnexpectedWeightKeyLength = fmt.Errorf("expected weight key length %d", weightKeyLength)
+	errUnexpectedBLSKeyLength    = fmt.Errorf("expected bls key length %d", blsKeyLength)
+)
+
+func getStartWeightKey(subnetID ids.ID, height uint64) []byte {
+	key := make([]byte, startWeightKeyLength)
+	copy(key, subnetID[:])
+	binary.BigEndian.PutUint64(key[weightKeyHeightOffset:], ^height)
+	return key
+}
+
+func getWeightKey(subnetID ids.ID, height uint64, nodeID ids.NodeID) []byte {
+	key := make([]byte, weightKeyLength)
+	copy(key, subnetID[:])
+	binary.BigEndian.PutUint64(key[weightKeyHeightOffset:], ^height)
+	copy(key[weightKeyNodeIDOffset:], nodeID[:])
+	return key
+}
+
+func parseWeightKey(key []byte) (ids.ID, uint64, ids.NodeID, error) {
+	if len(key) != weightKeyLength {
+		return ids.Empty, 0, ids.EmptyNodeID, errUnexpectedWeightKeyLength
+	}
+	var (
+		subnetID ids.ID
+		nodeID   ids.NodeID
```

Comment:
if we pass subnetID and nodeID among function parameters, we may be able to avoid these allocations. We'd let the caller allocate once and overwrite the variables as needed.
Not sure it'd be a big difference but we may consider it.